src/Libraries/Microsoft.Extensions.AI/ChatCompletion/ChatClientBuilder.cs

```diff
@@ -41,27 +46,61 @@ public ChatClientBuilder(Func<IServiceProvider, IChatClient> innerClientFactory)
-        if (_clientFactories is not null)
+        if (_clientMiddleware is not null)
         {
-            for (var i = _clientFactories.Count - 1; i >= 0; i--)
+            for (var i = _clientMiddleware.Count - 1; i >= 0; i--)
             {
-                chatClient = _clientFactories[i](chatClient, services);
-                if (chatClient is null)
+                chatClientFactory = _clientMiddleware[i](chatClientFactory);
+
+                if (chatClientFactory is null)
                 {
                     Throw.InvalidOperationException(
-                        $"The {nameof(ChatClientBuilder)} entry at index {i} returned null. " +
-                        $"Ensure that the callbacks passed to {nameof(Use)} return non-null {nameof(IChatClient)} instances.");
+                        $"The {nameof(ChatClientBuilder)} middleware entry at index {i} returned null. " +
+                        $"Ensure that the callbacks passed to {nameof(Use)} return non-null {nameof(Func<IFeatureCollection, IChatClient>)} instances.");
                 }
             }
         }
 
+        var features = new FeatureCollection();
+        services ??= EmptyServiceProvider.Instance;
+        var chatClient = chatClientFactory(features, services);
+
+        if (chatClient is null)
+        {
+            Throw.InvalidOperationException(
+                $"The {nameof(ChatClientBuilder)} middleware pipeline returned null." +
+                $"Ensure that the callbacks passed to {nameof(Use)} return non-null {nameof(IChatClient)} instances.");
+        }
+
         return chatClient;
     }
 
+    /// <summary>Adds a factory for an intermediate chat client to the chat client pipeline.</summary>
+    /// <param name="clientMiddleware">The client middleware function.</param>
+    /// <returns>The updated <see cref="ChatClientBuilder"/> instance.</returns>
+    /// <exception cref="ArgumentNullException"><paramref name="clientMiddleware"/> is <see langword="null"/>.</exception>
+    public ChatClientBuilder Use(Func<Func<IFeatureCollection, IChatClient>, IServiceProvider, Func<IFeatureCollection, IChatClient>> clientMiddleware)
+    {
+        _ = Throw.IfNull(clientMiddleware);
+
+        (_clientMiddleware ??= []).Add(nextFactory =>
+        {
+            // Pass the services around the user-defined middleware since we don't allow them to pass the IServiceProvider to the next() middleware.
+            // It'd be simpler to make passing the IServiceProvider to the next middleware explicit, but as before, it's probably best to
+            // not allow IServiceProvider to be changed from one middleware to the next.
+            return (features, services) =>
+            {
+                var appliedMiddleware = clientMiddleware(fc => nextFactory(fc, services), services);
+                return appliedMiddleware(features);
```

Comment:
What guarantees appliedMiddleware is not null?

Reply:
I think the check on line 58 still guards against the "appliedMiddleware" ClientFactory being null. What we currently don't check until the end of the pipeline is whether the returned IChatClient itself is null. I can go back to checking this at every stage.